Configurator "Output YAML" Flag

chases2 opened the pull request:
Added to allow conversion from prow annotations to only Testgrid YAML

In this way, other instances of Configurator could chain their outputs into a sharded Testgrid Config folder.
Revision of #13510 for issue #13455 

The prow job annotation functions now take a flag, but the flag is really doing two things:
- Allow generation of dashboards from prow annotations
- Suppress the "reconciliation" of the default variables
Should it be two booleans instead?

chases2 commented:
/cc @Katharine 
/cc @michelle192837 
/assign @shyamjvs

Katharine commented on testgrid/cmd/configurator/prow.go:
```diff
@@ -129,7 +131,16 @@ func applySingleProwjobAnnotations(c *Config, pc *prowConfig.Config, j prowConfi
 			d := c.config.FindDashboard(dashboardName)
 			if d == nil {
-				return fmt.Errorf("couldn't find dashboard %q for job %q", dashboardName, j.Name)
+				if !generateYAML {
```
I think we should insist on dashboards being defined in the same set of files as the things that go on them, and not let this pass without comment (and then potentially fail on our side, which will be more painful).

chases2 replied:
I think I'd rather keep the YAML config as the authoritative source of dashboards for now; it lets us keep the typo protection and lets us keep blending config and annotations.

I'm interested in additional controls on what's defined where, especially if it keeps things easy to understand, but I think that's beyond this PR.

Katharine replied:
I agree, but this code seems to have the opposite effect - if a dashboard doesn't exist, it will now just silently invent it and assume that's okay.

I'm not sure how you're envisioning running this, but it must involve some amount of yaml in the other repo to define their dashboards and whatnot too?

chases2 replied:
For clarity, I very recently pushed a revision that removes this particular change; now it throws an error if it sees a dashboard that isn't in the YAML like it used to.

I originally envisioned the program simply taking a series of prow jobs with annotations and, assuming they're perfectly accurate, constructing a yaml configuration file. In the [first draft](https://github.com/kubernetes/test-infra/pull/13510), I had to specifically specify a GCS bucket and a configuration file for all those defaults.

On the first pass through of this, I thought it would be neat to have it generate dashboards as well, so you only need a prow config and also a testgrid config that contains some defaults.

After this conversation, I _now_ think that's way more error-prone than is worthwhile, and think it's best with a prow config and also a testgrid config that contains defaults and dashboards.

Katharine commented on testgrid/default-template.yaml:
```diff
@@ -0,0 +1,38 @@
+# Default Settings are required for TestGrid YAML Configuration files
```
What is this file for?

chases2 replied:
It serves two purposes:
1) The specific use case I'm writing for, pulling annotations into a YAML file, still requires default settings. Both Configurator and the standalone solution would set those default settings by pulling them from a document like this.
2) Documentation. If there's going to be no single "config.yaml" with a bunch of comments at the top describing what all the defaults are, we should maintain a copy-pastable template that describes what all those defaults are.

If we insist on dashboards being defined in the YAML, then I would add to this template something like `# here's an example dashboard and all the dashboards you want need to go here`

Katharine replied:
1. That makes sense and was my originaly assumption, but it doesn't seem to actually be referenced anywhere? Or am I just blind?

chases2 replied:
I'm not pointing it out anywhere in this PR explicitly.

Updating the READMEs with how this and sharding change TestGrid workflows is a TODO in my near future. Is there something I could change to make this easier to discover?

Katharine replied:
On further thought I think this is fine for now.

chases2 commented:
I want to keep the defaulting out with YAML to make the YAML considerably smaller. Using Prow's annotations as an example, keeping the "reconcile" calls out is the difference between ~7,000 lines and ~133,000 lines.

Katharine commented:
Fair enough.

Katharine commented on testgrid/cmd/configurator/main.go:
```diff
@@ -25,6 +25,7 @@ import (
 	"log"
 	"net/url"
 	"os"
+	"sigs.k8s.io/yaml"
```
This should be in the second import block, not this one.

Katharine commented on testgrid/default-template.yaml:
```diff
@@ -0,0 +1,46 @@
+
+dashboards:
+  - name: example-dashboard-one
+  - name: example-dashboard-two
```
Trailing newline!

chases2 commented:
I reworked that error a bit to show all the many things you can do with Configurator now. Pushing to stdout sounds like a pretty useful feature to maintain, so I added it for YAML also
